Firestore/core/include/firebase/firestore/timestamp.h

```diff
@@ -0,0 +1,173 @@
+   * `std::chrono::system_clock`. Otherwise, the timestamp will have second
+   * precision.
+   *
+   * @return a new timestamp representing the current date.
+   */
+  static Timestamp Now();
+
+  /**
+   * The number of seconds of UTC time since Unix epoch 1970-01-01T00:00:00Z.
+   */
+  std::int64_t seconds() const {
+    return seconds_;
+  }
+
+  /**
+   * The non-negative fractions of a second at nanosecond resolution. Negative
+   * second values with fractions still have non-negative nanoseconds values
+   * that count forward in time.
+   */
+  std::int32_t nanoseconds() const {
+    return nanoseconds_;
+  }
+
+  /**
+   * Converts `std::time_t` to a Timestamp.
+   *
+   * @param seconds_since_unix_epoch The number of seconds of UTC time since
+   *     Unix epoch 1970-01-01T00:00:00Z. Can be negative to represent dates
+   *     before the epoch.
+   *
+   *     Note that while the epoch of `std::time_t` is unspecified, it's usually
+   *     Unix epoch. If this assumption is broken, this function will produce
+   *     incorrect results.
+   *
+   * @return a new timestamp with the given number of seconds and zero
+   *     nanoseconds.
+   */
+  static Timestamp FromTimeT(std::time_t seconds_since_unix_epoch);
+
+#if !defined(_STLPORT_VERSION)
+  /**
+   * Converts `std::chrono::time_point` to a Timestamp.
+   *
+   * @param time_point The time point with system clock's epoch, which is
+   *     presumed to be Unix epoch 1970-01-01T00:00:00Z. Can be negative to
+   *     represent dates before the epoch.
+   *
+   *     Note that while the epoch of `std::chrono::system_clock` is
+   *     unspecified, it's usually Unix epoch. If this assumption is broken,
+   *     this constructor will produce incorrect results.
+   */
+  Timestamp(std::chrono::time_point<std::chrono::system_clock> time_point);
+#endif
+
+ private:
+  // Checks that the number of seconds is within the supported date range, and
+  // that nanoseconds satisfy 0 <= ns <= 1second.
+  void ValidateBounds();
+
+  std::int64_t seconds_ = 0;
+  std::int32_t nanoseconds_ = 0;
+};
+
+inline bool operator<(const Timestamp& lhs, const Timestamp& rhs) {
+  return lhs.seconds() < rhs.seconds() ||
+         (lhs.seconds() == rhs.seconds() &&
+          lhs.nanoseconds() < rhs.nanoseconds());
+}
+
+inline bool operator>(const Timestamp& lhs, const Timestamp& rhs) {
+  return rhs < lhs;
+}
+
+inline bool operator>=(const Timestamp& lhs, const Timestamp& rhs) {
+  return !(lhs < rhs);
+}
+
+inline bool operator<=(const Timestamp& lhs, const Timestamp& rhs) {
+  return !(lhs > rhs);
+}
+
+inline bool operator!=(const Timestamp& lhs, const Timestamp& rhs) {
+  return lhs < rhs || lhs > rhs;
+}
+
+inline bool operator==(const Timestamp& lhs, const Timestamp& rhs) {
+  return !(lhs != rhs);
+}
+
+}  // namespace firebase
+
+namespace std {
+template <>
+struct hash<firebase::Timestamp> {
+  size_t operator()(const firebase::Timestamp& timestamp) const {
```

Comment:
This is adapted from the Java version (which, in turn, seems to be based on the approach from "Effective Java").

Reply:
Right, in Java, hash has to be `int`, which, IIRC, is 32 bits.

Now that I think about it, it seems that the initial value doesn't matter at all and might as well be zero (I think it only matters for arrays), so it becomes `(0*37 + seconds) * 37 + nanoseconds = 37 * seconds + nanoseconds`. Correct me if I'm wrong.

It would be great to have some proper hash functions, but `Timestamp` being intended to move into `common` probably won't be able to use them anyway.